main.cpp

```diff
@@ -39,25 +39,355 @@ Purpose:  This project continues developing Project3.
+}
+
+void Computer::eraseDisk(std::string volumeName)
+{
+    std::cout << "Erased Disk " << volumeName << std::endl;
+}
+
+void Computer::showInfo()
+{
+    std::cout << "The Computer has " << amountOfRam << "GB of RAM and a processor speed of " << processorSpeed << " GHz" << std::endl;
+}
+
+void Computer::powerOffCountdown() 
+{
+    int i = 10;
+    while (i > 0)
+    {
+        std::cout << "Power off in " << i << std::endl;
+        --i;
+    }
+    shutDown();
+} 
+
+void Computer::Application::start()
+{
+    std::cout << Computer::Application::name << " started" << std::endl;
+}
+
+void Computer::Application::close()
+{
+    std::cout << Computer::Application::name << " closed" << std::endl;
+}
+
+void Computer::Application::install(std::string filePath)
+{
+    std::cout << Computer::Application::name << " installed on " << filePath << std::endl;
+}
+// ============================================================
 
 /*
  copied UDT 3:
  */
+struct Bus
+{
+    Bus();
+    ~Bus();
+    
+    int maximumSpeed { 110 };
+    int numberOfSeats;
+    std::string manufacturer = "MAN";
+    float fuelConsumption = 20.3f;
+    float fuelCapacity { 200.0f };
+
+    void startEngine();
+    void turnLeft(int angle);
+    void openDoors(bool openAllDoors);
+    void drive (int distance);
+};
+
+Bus::Bus() : 
+numberOfSeats(24) 
+{
+    std::cout << "CONSTRUCTING Bus" << std::endl;
+}
+
+Bus::~Bus() 
+{
+    std::cout << "DESTRUCTING Bus" << std::endl;
+}
+
+void Bus::startEngine()
+{
+    std::cout << "Engine of Bus started. The maximum speed is " << maximumSpeed << " km/h" <<std::endl;
+}
+
+void Bus::turnLeft(int angle)
+{
+    std::cout << "Turned left " << angle << " degrees" << std::endl;
+}
+
+void Bus::openDoors(bool openAllDoors)
+{
+    if(openAllDoors)
+    {
+        std::cout << "Opened all doors" << std::endl;
+    }
+    else 
+    {
+        std::cout << "Opened a single door" << std::endl;
+    }
+}
+
+void Bus::drive (int distance) 
+{
+    float fuelLeft, fuelUsed;
+    int i = 0; 
+    while ( i <= distance )
+    {
+        fuelUsed =  i * fuelConsumption / 100;
+        fuelLeft =  fuelCapacity - fuelUsed;
+
+        if ( fuelLeft < 0 ) 
+        {
+            std::cout << "YOU RAN OUT OF FUEL!!!" << std::endl;
+            break;
+        }
+
+        std::cout << "You have driven " << i << " Km and used " << fuelUsed << " liters of fuel. You have " << fuelLeft << " liters left." << std::endl;
+        i += 50;        
+    }
+}
+// ============================================================
 
 /*
  new UDT 4:
  with 2 member functions
  */
+struct musicStore
+{
+    musicStore();
+    ~musicStore();
+
+    ElectricGuitar lesPaul { "Gibson", "Les Paul", 3289.99f };
+    ElectricGuitar stratocaster { "Fender", "Stratocaster", 2399.99f };
+    Computer officePc { "Asus", "Windows 95" };
+
+    float valueOfGuitars();
+    float blackFridayPrice(int discount, float originalPrice);
```

Comment:
`blackFridayPrice` is a noun, not an action.
functions are actions.
- [ ] revise